README.md
# Koverto - UUID

[![Go](https://github.com/koverto/uuid/workflows/Go/badge.svg)][workflow]
[![Go Report Card](https://goreportcard.com/badge/github.com/koverto/uuid)](https://goreportcard.com/report/github.com/koverto/uuid)

Wraps [github.com/google/uuid][uuid] for use as a [protobuf][] type and with
implementations of various serialization interfaces.

## Usage

```protobuf
import "github.com/koverto/uuid/uuid.proto";

message User {
    uuid.UUID id = 1;
}
```

## Copyright

Copyright © 2020 Jesse B. Hannah. Licensed under the [GNU GPL version 3 or
later][gpl].

[gpl]: LICENSE
[protobuf]: https://developers.google.com/protocol-buffers/
[uuid]: https://github.com/google/uuid
[workflow]: https://github.com/koverto/uuid/actions?query=workflow%3AGo


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][], and this project adheres to
[Semantic Versioning][].

## Unreleased

## v1.3.0 - 2020-03-04

### Added

- `func Parse(s string) (*UUID, error)`

### Changed

- Updated go.mongodb.org/mongo-driver from 1.3.0 to 1.3.1

## v1.2.1 - 2020-02-17

### Fixed

- Implement JSON and GraphQL un/marshaling on the correct struct

## v1.2.0 - 2020-02-17

### Added

- Added JSON support
- Added GraphQL support

## v1.1.0 - 2020-02-12

### Added

- Added BSON support

## v1.0.0 - 2020-02-12

### Added

- `UUID` protobuf message type

[keep a changelog]: https://keepachangelog.com/en/1.0.0/
[semantic versioning]: https://semver.org/spec/v2.0.0.html


uuid_test.go
package uuid

import (
	fmt "fmt"
	"reflect"
	"testing"

	"github.com/google/uuid"
	"go.mongodb.org/mongo-driver/bson/bsontype"
)

const uuidString = "4915e120-3594-b29e-bd92-62abff23e1c6"

func uuidBytes() [16]byte {
	return [16]byte{0x49, 0x15, 0xe1, 0x20, 0x35, 0x94, 0xb2, 0x9e, 0xbd, 0x92, 0x62, 0xab, 0xff, 0x23, 0xe1, 0xc6}
}

func uuidZero() [16]byte {
	return [16]byte{0x0}
}

func TestNew(t *testing.T) {
	tests := []struct {
		name string
		want int
	}{
		{"New UUID", 16},
	}
	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := New()
			if got := u.Uuid.Size(); !reflect.DeepEqual(got, test.want) {
				t.Errorf("New().Size() = %v, want %v", got, test.want)
			}
		})
	}
}

func TestUUID_Size(t *testing.T) {
	uuidBytes := uuidBytes()
	tests := []struct {
		name   string
		fields *_uuid
		want   int
	}{
		{"Valid UUID", &_uuid{uuid.UUID(uuidBytes)}, 16},
		{"Nil UUID", nil, 0},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			if got := test.fields.Size(); got != test.want {
				t.Errorf("_uuid.Size() = %v, want %v", got, test.want)
			}
		})
	}
}

func TestUUID_MarshalTo(t *testing.T) {
	uuidBytes := uuidBytes()
	tests := []struct {
		name    string
		fields  *_uuid
		args    []byte
		want    int
		wantErr bool
	}{
		{"Valid UUID", &_uuid{uuid.UUID(uuidBytes)}, make([]byte, 0), 16, false},
		{"Nil UUID", nil, make([]byte, 0), 0, false},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			got, err := test.fields.MarshalTo(test.args)
			if (err != nil) != test.wantErr {
				t.Errorf("_uuid.MarshalTo() error = %v, wantErr %v", err, test.wantErr)
				return
			}
			if got != test.want {
				t.Errorf("_uuid.MarshalTo() = %v, want %v", got, test.want)
			}
		})
	}
}

func TestUUID_Unmarshal(t *testing.T) {
	uuidBytes := uuidBytes()
	tests := []struct {
		name    string
		args    []byte
		want    [16]byte
		wantErr bool
	}{
		{"Valid UUID", uuidBytes[:], uuidBytes, false},
		{"Zero length data", make([]byte, 0), [16]byte{}, false},
		{"Invalid data", []byte{0x00}, [16]byte{}, true},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := &_uuid{}
			if err := u.Unmarshal(test.args); (err != nil) != test.wantErr {
				t.Errorf("_uuid.Unmarshal() error = %v, wantErr %v", err, test.wantErr)
			}
			if u.UUID != test.want {
				t.Errorf("_uuid.Unmarshal() = %v, want %v", u.UUID, test.want)
			}
		})
	}
}

func TestUUID_MarshalBSONValue(t *testing.T) {
	uuidBytes := uuidBytes()
	header := []byte{0x10, 0x00, 0x00, 0x00, 0x04}

	tests := []struct {
		name    string
		want    bsontype.Type
		wantErr bool
		want1   []byte
		fields  uuid.UUID
	}{
		{"Valid UUID", bsontype.Binary, false, append(header, uuidBytes[:]...), uuidBytes},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := _uuid{
				UUID: test.fields,
			}
			got, got1, err := u.MarshalBSONValue()
			if (err != nil) != test.wantErr {
				t.Errorf("_uuid.MarshalBSONValue() error = %v, wantErr %v", err, test.wantErr)
				return
			}
			if !reflect.DeepEqual(got, test.want) {
				t.Errorf("_uuid.MarshalBSONValue() got = %v, want %v", got, test.want)
			}
			if !reflect.DeepEqual(got1, test.want1) {
				t.Errorf("_uuid.MarshalBSONValue() got1 = %v, want %v", got1, test.want1)
			}
		})
	}
}

func TestUUID_UnmarshalBSONValue(t *testing.T) {
	uuidBytes := uuidBytes()
	goodHeader := []byte{0x10, 0x00, 0x00, 0x00, 0x04}
	badTypeHeader := []byte{0x01, 0x00, 0x00, 0x00, 0x04}
	badSubtypeHeader := []byte{0x10, 0x00, 0x00, 0x00, 0x01}

	type args struct {
		bsonType bsontype.Type
		data     []byte
	}

	tests := []struct {
		name    string
		args    args
		want    [16]byte
		wantErr bool
	}{
		{"Good header", args{bsontype.Binary, append(goodHeader, uuidBytes[:]...)}, uuidBytes, false},
		{"Bad type", args{bsontype.Boolean, append(goodHeader, uuidBytes[:]...)}, [16]byte{}, true},
		{"Bad type header", args{bsontype.Binary, append(badTypeHeader, uuidBytes[:]...)}, [16]byte{}, true},
		{"Bad subtype header", args{bsontype.Binary, append(badSubtypeHeader, uuidBytes[:]...)}, [16]byte{}, true},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := &_uuid{}
			if err := u.UnmarshalBSONValue(test.args.bsonType, test.args.data); (err != nil) != test.wantErr {
				t.Errorf("_uuid.UnmarshalBSONValue() error = %v, wantErr %v", err, test.wantErr)
			}
		})
	}
}

func TestUUID_MarshalJSON(t *testing.T) {
	uuidBytes := uuidBytes()
	tests := []struct {
		name    string
		wantErr bool
		fields  uuid.UUID
		want    []byte
	}{
		{"Valid UUID", false, uuid.UUID(uuidBytes), []byte(fmt.Sprintf("\"%s\"", uuidString))},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := UUID{&_uuid{
				UUID: test.fields,
			}}
			got, err := u.MarshalJSON()
			if (err != nil) != test.wantErr {
				t.Errorf("UUID.MarshalJSON() error = %v, wantErr %v", err, test.wantErr)
				return
			}
			if !reflect.DeepEqual(got, test.want) {
				t.Errorf("UUID.MarshalJSON() = %v, want %v", got, test.want)
			}
		})
	}
}

func TestUUID_UnmarshalJSON(t *testing.T) {
	uuidBytes := uuidBytes()
	tests := []struct {
		name    string
		fields  uuid.UUID
		args    []byte
		wantErr bool
	}{
		{"Valid UUID", uuid.UUID(uuidBytes), []byte(uuidString), false},
		{"Invalid UUID", uuid.UUID{}, []byte(""), true},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			u := UUID{&_uuid{
				UUID: test.fields,
			}}
			if err := u.UnmarshalJSON(test.args); (err != nil) != test.wantErr {
				t.Errorf("UUID.UnmarshalJSON() error = %v, wantErr %v", err, test.wantErr)
			}
		})
	}
}

func TestParse(t *testing.T) {
	uuidBytes := uuidBytes()
	uuidZero := uuidZero()
	tests := []struct {
		name    string
		args    string
		want    *UUID
		wantErr bool
	}{
		{"Valid UUID", uuidString, &UUID{&_uuid{uuid.UUID(uuidBytes)}}, false},
		{"Invalid UUID", "invalid", &UUID{&_uuid{uuid.UUID(uuidZero)}}, true},
	}

	for _, tt := range tests {
		test := tt
		t.Run(test.name, func(t *testing.T) {
			got, err := Parse(test.args)
			if (err != nil) != test.wantErr {
				t.Errorf("Parse() error = %v, wantErr %v", err, test.wantErr)
				return
			}
			if !reflect.DeepEqual(got, test.want) {
				t.Errorf("Parse() = %v, want %v", got, test.want)
			}
		})
	}
}


uuid.go
//go:generate protoc --gogofaster_out=plugins=grpc:. --proto_path=$GOPATH/pkg/mod:. uuid.proto

// Package uuid wraps github.com/google/uuid for use as a protobuf type and with
// implementations of various de/serialization interfaces.
package uuid

import (
	fmt "fmt"
	io "io"
	"strconv"

	"github.com/google/uuid"
	"github.com/prometheus/common/log"
	"go.mongodb.org/mongo-driver/bson/bsontype"
	"go.mongodb.org/mongo-driver/x/bsonx"
)

// Parse parses a UUID from a valid string representation.
func Parse(s string) (*UUID, error) {
	parsed, err := uuid.Parse(s)
	return &UUID{&_uuid{parsed}}, err
}

// New generates a new, random UUID.
func New() *UUID {
	id := _uuid{uuid.New()}
	return &UUID{&id}
}

type _uuid struct {
	uuid.UUID
}

// Size is required to implement the proto.Marshaler interface.
func (u *_uuid) Size() int {
	if u == nil {
		return 0
	}

	return len(u.UUID)
}

// MarshalTo is required to implement the proto.Marshaler interface.
func (u *_uuid) MarshalTo(data []byte) (int, error) {
	if u == nil {
		return 0, nil
	}

	copy(data, u.UUID[:])

	return len(u.UUID), nil
}

// Unmarshal is required to implement the proto.Marshaler interface.
func (u *_uuid) Unmarshal(data []byte) error {
	if len(data) == 0 {
		u = nil
		return nil
	}

	uid, err := uuid.FromBytes(data)
	if err == nil {
		u.UUID = uid
	}

	return err
}

// MarshalBSONValue implements the bson.ValueMarshaler interface.
func (u _uuid) MarshalBSONValue() (bsontype.Type, []byte, error) {
	val := bsonx.Binary(bsontype.BinaryUUID, u.UUID[:])
	return val.MarshalBSONValue()
}

// UnmarshalBSONValue implements the bson.ValueUnmarshaler interface.
func (u *_uuid) UnmarshalBSONValue(bsonType bsontype.Type, data []byte) error {
	if bsonType != bsontype.Binary || data[0] != 0x10 || data[4] != bsontype.BinaryUUID {
		return fmt.Errorf("could not unmarshal %v as a UUID", bsonType)
	}

	return u.Unmarshal(data[5:])
}

// UnmarshalGQL implements the graphql.Unmarshal interface.
func (u *UUID) UnmarshalGQL(v interface{}) error {
	str, ok := v.(string)
	if !ok {
		return fmt.Errorf("value for unmarshalling was not a string: %v", v)
	}

	return u.UnmarshalJSON([]byte(str))
}

// MarshalGQL implements the graphql.Marshal interface.
func (u UUID) MarshalGQL(w io.Writer) {
	marshaled, _ := u.MarshalJSON()
	if _, err := w.Write(marshaled); err != nil {
		log.Errorf("Error marshalling %v to GraphQL: %s", u, err)
	}
}

// MarshalJSON implements the json.Marshaler interface.
func (u UUID) MarshalJSON() ([]byte, error) {
	return []byte(strconv.Quote(u.Uuid.String())), nil
}

// UnmarshalJSON implements the json.Unmarshaler interface.
func (u *UUID) UnmarshalJSON(data []byte) error {
	if parsed, err := uuid.Parse(string(data)); err == nil {
		u.Uuid = &_uuid{parsed}
	} else {
		return err
	}

	return nil
}


uuid.pb.go
// Code generated by protoc-gen-gogo. DO NOT EDIT.
// source: uuid.proto

package uuid

import (
	fmt "fmt"
	_ "github.com/gogo/protobuf/gogoproto"
	proto "github.com/gogo/protobuf/proto"
	io "io"
	math "math"
)

// Reference imports to suppress errors if they are not otherwise used.
var _ = proto.Marshal
var _ = fmt.Errorf
var _ = math.Inf

// This is a compile-time assertion to ensure that this generated file
// is compatible with the proto package it is being compiled against.
// A compilation error at this line likely means your copy of the
// proto package needs to be updated.
const _ = proto.GoGoProtoPackageIsVersion2 // please upgrade the proto package

type UUID struct {
	Uuid *_uuid `protobuf:"bytes,1,opt,name=uuid,proto3,customtype=_uuid" json:"uuid,omitempty"`
}

func (m *UUID) Reset()         { *m = UUID{} }
func (m *UUID) String() string { return proto.CompactTextString(m) }
func (*UUID) ProtoMessage()    {}
func (*UUID) Descriptor() ([]byte, []int) {
	return fileDescriptor_2b0c78c62e9a4add, []int{0}
}
func (m *UUID) XXX_Unmarshal(b []byte) error {
	return m.Unmarshal(b)
}
func (m *UUID) XXX_Marshal(b []byte, deterministic bool) ([]byte, error) {
	if deterministic {
		return xxx_messageInfo_UUID.Marshal(b, m, deterministic)
	} else {
		b = b[:cap(b)]
		n, err := m.MarshalTo(b)
		if err != nil {
			return nil, err
		}
		return b[:n], nil
	}
}
func (m *UUID) XXX_Merge(src proto.Message) {
	xxx_messageInfo_UUID.Merge(m, src)
}
func (m *UUID) XXX_Size() int {
	return m.Size()
}
func (m *UUID) XXX_DiscardUnknown() {
	xxx_messageInfo_UUID.DiscardUnknown(m)
}

var xxx_messageInfo_UUID proto.InternalMessageInfo

func init() {
	proto.RegisterType((*UUID)(nil), "uuid.UUID")
}

func init() { proto.RegisterFile("uuid.proto", fileDescriptor_2b0c78c62e9a4add) }

var fileDescriptor_2b0c78c62e9a4add = []byte{
	// 141 bytes of a gzipped FileDescriptorProto
	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0xff, 0xe2, 0xe2, 0x2a, 0x2d, 0xcd, 0x4c,
	0xd1, 0x2b, 0x28, 0xca, 0x2f, 0xc9, 0x17, 0x62, 0x01, 0xb1, 0xa5, 0x4c, 0xd2, 0x33, 0x4b, 0x32,
	0x4a, 0x93, 0xf4, 0x92, 0xf3, 0x73, 0xf5, 0xd3, 0xf3, 0xd3, 0xf3, 0xf5, 0xc1, 0x92, 0x49, 0xa5,
	0x69, 0x0e, 0x65, 0x86, 0x7a, 0xc6, 0x7a, 0x86, 0x60, 0x41, 0xb0, 0x18, 0x98, 0x05, 0xd1, 0xab,
	0xa4, 0xca, 0xc5, 0x12, 0x1a, 0xea, 0xe9, 0x22, 0x24, 0xcb, 0x05, 0x36, 0x45, 0x82, 0x51, 0x81,
	0x51, 0x83, 0xc7, 0x89, 0xf3, 0xd6, 0x3d, 0x79, 0xd6, 0x78, 0x90, 0x40, 0x10, 0x58, 0xd8, 0x49,
	0xe2, 0xc4, 0x23, 0x39, 0xc6, 0x0b, 0x8f, 0xe4, 0x18, 0x1f, 0x3c, 0x92, 0x63, 0x9c, 0xf0, 0x58,
	0x8e, 0xe1, 0xc2, 0x63, 0x39, 0x86, 0x1b, 0x8f, 0xe5, 0x18, 0x92, 0xd8, 0xc0, 0xe6, 0x18, 0x03,
	0x02, 0x00, 0x00, 0xff, 0xff, 0xfb, 0xec, 0x02, 0xb7, 0x91, 0x00, 0x00, 0x00,
}

func (m *UUID) Marshal() (dAtA []byte, err error) {
	size := m.Size()
	dAtA = make([]byte, size)
	n, err := m.MarshalTo(dAtA)
	if err != nil {
		return nil, err
	}
	return dAtA[:n], nil
}

func (m *UUID) MarshalTo(dAtA []byte) (int, error) {
	var i int
	_ = i
	var l int
	_ = l
	if m.Uuid != nil {
		dAtA[i] = 0xa
		i++
		i = encodeVarintUuid(dAtA, i, uint64(m.Uuid.Size()))
		n1, err := m.Uuid.MarshalTo(dAtA[i:])
		if err != nil {
			return 0, err
		}
		i += n1
	}
	return i, nil
}

func encodeVarintUuid(dAtA []byte, offset int, v uint64) int {
	for v >= 1<<7 {
		dAtA[offset] = uint8(v&0x7f | 0x80)
		v >>= 7
		offset++
	}
	dAtA[offset] = uint8(v)
	return offset + 1
}
func (m *UUID) Size() (n int) {
	if m == nil {
		return 0
	}
	var l int
	_ = l
	if m.Uuid != nil {
		l = m.Uuid.Size()
		n += 1 + l + sovUuid(uint64(l))
	}
	return n
}

func sovUuid(x uint64) (n int) {
	for {
		n++
		x >>= 7
		if x == 0 {
			break
		}
	}
	return n
}
func sozUuid(x uint64) (n int) {
	return sovUuid(uint64((x << 1) ^ uint64((int64(x) >> 63))))
}
func (m *UUID) Unmarshal(dAtA []byte) error {
	l := len(dAtA)
	iNdEx := 0
	for iNdEx < l {
		preIndex := iNdEx
		var wire uint64
		for shift := uint(0); ; shift += 7 {
			if shift >= 64 {
				return ErrIntOverflowUuid
			}
			if iNdEx >= l {
				return io.ErrUnexpectedEOF
			}
			b := dAtA[iNdEx]
			iNdEx++
			wire |= (uint64(b) & 0x7F) << shift
			if b < 0x80 {
				break
			}
		}
		fieldNum := int32(wire >> 3)
		wireType := int(wire & 0x7)
		if wireType == 4 {
			return fmt.Errorf("proto: UUID: wiretype end group for non-group")
		}
		if fieldNum <= 0 {
			return fmt.Errorf("proto: UUID: illegal tag %d (wire type %d)", fieldNum, wire)
		}
		switch fieldNum {
		case 1:
			if wireType != 2 {
				return fmt.Errorf("proto: wrong wireType = %d for field Uuid", wireType)
			}
			var byteLen int
			for shift := uint(0); ; shift += 7 {
				if shift >= 64 {
					return ErrIntOverflowUuid
				}
				if iNdEx >= l {
					return io.ErrUnexpectedEOF
				}
				b := dAtA[iNdEx]
				iNdEx++
				byteLen |= (int(b) & 0x7F) << shift
				if b < 0x80 {
					break
				}
			}
			if byteLen < 0 {
				return ErrInvalidLengthUuid
			}
			postIndex := iNdEx + byteLen
			if postIndex > l {
				return io.ErrUnexpectedEOF
			}
			var v _uuid
			m.Uuid = &v
			if err := m.Uuid.Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
				return err
			}
			iNdEx = postIndex
		default:
			iNdEx = preIndex
			skippy, err := skipUuid(dAtA[iNdEx:])
			if err != nil {
				return err
			}
			if skippy < 0 {
				return ErrInvalidLengthUuid
			}
			if (iNdEx + skippy) > l {
				return io.ErrUnexpectedEOF
			}
			iNdEx += skippy
		}
	}

	if iNdEx > l {
		return io.ErrUnexpectedEOF
	}
	return nil
}
func skipUuid(dAtA []byte) (n int, err error) {
	l := len(dAtA)
	iNdEx := 0
	for iNdEx < l {
		var wire uint64
		for shift := uint(0); ; shift += 7 {
			if shift >= 64 {
				return 0, ErrIntOverflowUuid
			}
			if iNdEx >= l {
				return 0, io.ErrUnexpectedEOF
			}
			b := dAtA[iNdEx]
			iNdEx++
			wire |= (uint64(b) & 0x7F) << shift
			if b < 0x80 {
				break
			}
		}
		wireType := int(wire & 0x7)
		switch wireType {
		case 0:
			for shift := uint(0); ; shift += 7 {
				if shift >= 64 {
					return 0, ErrIntOverflowUuid
				}
				if iNdEx >= l {
					return 0, io.ErrUnexpectedEOF
				}
				iNdEx++
				if dAtA[iNdEx-1] < 0x80 {
					break
				}
			}
			return iNdEx, nil
		case 1:
			iNdEx += 8
			return iNdEx, nil
		case 2:
			var length int
			for shift := uint(0); ; shift += 7 {
				if shift >= 64 {
					return 0, ErrIntOverflowUuid
				}
				if iNdEx >= l {
					return 0, io.ErrUnexpectedEOF
				}
				b := dAtA[iNdEx]
				iNdEx++
				length |= (int(b) & 0x7F) << shift
				if b < 0x80 {
					break
				}
			}
			iNdEx += length
			if length < 0 {
				return 0, ErrInvalidLengthUuid
			}
			return iNdEx, nil
		case 3:
			for {
				var innerWire uint64
				var start int = iNdEx
				for shift := uint(0); ; shift += 7 {
					if shift >= 64 {
						return 0, ErrIntOverflowUuid
					}
					if iNdEx >= l {
						return 0, io.ErrUnexpectedEOF
					}
					b := dAtA[iNdEx]
					iNdEx++
					innerWire |= (uint64(b) & 0x7F) << shift
					if b < 0x80 {
						break
					}
				}
				innerWireType := int(innerWire & 0x7)
				if innerWireType == 4 {
					break
				}
				next, err := skipUuid(dAtA[start:])
				if err != nil {
					return 0, err
				}
				iNdEx = start + next
			}
			return iNdEx, nil
		case 4:
			return iNdEx, nil
		case 5:
			iNdEx += 4
			return iNdEx, nil
		default:
			return 0, fmt.Errorf("proto: illegal wireType %d", wireType)
		}
	}
	panic("unreachable")
}

var (
	ErrInvalidLengthUuid = fmt.Errorf("proto: negative length found during unmarshaling")
	ErrIntOverflowUuid   = fmt.Errorf("proto: integer overflow")
)
